File: LinkedList/TestLinkedList.js
class Node{
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

class LinkedList {
    constructor() {
        this.head = null;
    }

    append(data) {
        const node = new Node(data);
        if (!this.head) {
            this.head = node;
        } else {
            let temp = this.head;
            while(temp.next) {
                temp = temp.next;
            }
            temp.next = node;
        }
    }

    reverse() {
        let new_node = null;
        let temp = this.head;
        while (temp) {
            let next_node = temp.next;
            temp.next = new_node;
            new_node = temp;
            temp = next_node;
        }
        return new_node;
    }

    print() {
        let node = this.head;
        while(node.next){
            console.log(node.data);
            node = node.next;
        }
    }

    printBackwards() {
        let node = this.reverse()
        while(node.next){
            console.log(node.data);
            node = node.next;
        }
    }

}

const ll = new LinkedList();

for (let i = 0; i <= 5; i++) {
    ll.append(i);
}

ll.print();
console.log()
ll.printBackwards();

File: LinkedList/TestLinkedList.go
package main

import "fmt"

type Node struct {
	data interface{}
	next *Node
}

type LinkedList struct {
	head *Node
}

func (list *LinkedList) append(data interface{}) {
	var node *Node = &Node{
		data: data,
		next: nil,
	}

	if list.head == nil {
		list.head = node
	} else {
		var temp *Node = list.head
		for temp.next != nil {
			temp = temp.next
		}
		temp.next = node
	}
}

func (list *LinkedList) reverse() (newNode *Node){
	// use bubble sort for reversing
	newNode = nil
	var temp *Node = list.head
	for temp != nil {
		var next_node *Node = temp.next
		temp.next = newNode
		newNode = temp
		temp = next_node
	}
	return
}

func (list *LinkedList) print() {
	var node *Node = list.head
	for node != nil {
		fmt.Println(node.data)
		node = node.next
	}
}

func (list *LinkedList) printBackwards() {
	var node *Node = list.reverse()
	for node != nil {
		fmt.Println(node.data)
		node = node.next
	}
}

func main() {
	var ll *LinkedList = &LinkedList{}

	for i := 0; i <= 5; i++ {
		ll.append(i)
	}

	ll.print()
	fmt.Println()
	ll.printBackwards()

}